src/util/handlers/PrimaryHandler.ts

```diff
@@ -0,0 +1,27 @@
+import type { ClusterManager } from '../../init/ClusterManager';
+import { InternalServerError } from '../errors/InternalServerError';
+import type { AsyncHandler } from './AsyncHandler';
+import { SequenceHandler } from './SequenceHandler';
+
+/**
+ * A composite handler that will only run all supporting handlers sequentially
+ * if they are executed from the primary process and return the value of the
+ * last supported handler.
+ * The `canHandle` check of this handler will always succeed if single threaded,
+ * or if run from the primary process.
+ */
+export class PrimaryHandler<TIn = void, TOut = void> extends SequenceHandler<TIn, TOut | undefined> {
```

Comment:
While this is going to make the config a bit more verbose, I would have this be a base utility handler extending `AsyncHandler` instead of `SequenceHandler`. It can then wrap around the `SequenceHandler`.

Reply:
You mean in the same way that a ConditionalHandler or a BooleanHandlers wraps around another/other handler(s)?

Reply:
```ts
export class PrimaryHandler<TIn, TOut> extends AsyncHandler<TIn, TOut> {
  private readonly clusterManager: ClusterManager;
  private readonly source: AsyncHandler<TIn, TOut>;

  public constructor(source: AsyncHandler<TIn, TOut>, clusterManager: ClusterManager) {
    super();
    this.source = source;
    this.clusterManager = clusterManager;
  }

  public async canHandle(input: TIn): Promise<void> {
    await this.checkPrimary();
    await this.source.canHandle(input);
  }

  public async handleSafe(input: TIn): Promise<TOut> {
    await this.checkPrimary();
    return this.source.handleSafe(input);
  }

  public async handle(input: TIn): Promise<TOut> {
    return this.source.handle(input);
  }

  /**
   * Checks if the condition has already been fulfilled.
   */
  private async checkPrimary(): Promise<void> {
    if (!this.clusterManager.isSingleThreadedOrPrimary()) {
      throw new NotImplementedHttpError('Will not execute on worker process.');
    }
  }
}
```

Reply:
Yes exactly